docs: scope functions (#1317)
JetBrains/kotlin-web-site: pages/docs/reference/scope-functions.md

<div class="sample" markdown="1" theme="idea">
```kotlin
data class Person(var name: String, var age: Int = 0, var city: String = "")

fun main() {
//sampleStart
    val adam = Person("Adam").apply { 
        age = 20                       // same as this.age = 20 or adam.age = 20
        city = "London"
    }
//sampleEnd
}
```
</div>

#### it

In turn, `let` and `also` have the context object as a lambda argument. If the argument name is not specified, the object is accessed by the implicit default name `it`. `it` is shorter than `this` and expressions with `it` are usually easier for reading. However, you can't omit the argument name, so there is no way to shorten calls to the object functions or properties. Hence, having the context object as `it` is better when the object is mostly used as an argument in function calls. `it` is also better if you operate multiple variables in the code block.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun writeToLog(str: String) {}
fun writeToFile(str: String) {}

fun main() {
//sampleStart

>     "Hello".also {

Author: Replaced the example

        println("Also called on the string $it")
        writeToLog(it)
        writeToFile(it)
    }
//sampleEnd
}
```
</div>

Additionally, when you pass the context object as an argument, you can provide a custom name for the context object inside the scope.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun writeToLog(str: String) {}

fun main() {
//sampleStart
    "Hello".also { myString -> 
        println("Also called on the string $myString")
        writeToLog(myString)
    }
//sampleEnd
}
```
</div>

### Return value

Scope functions return different values:
* `apply` and `also` return the context object.
* `let`, `run`, and `with` return the lambda result.

These two options let you choose the proper function depending on what you do next in your code.

> Having the object itself as a return value is useful for chaining other operations on it.

Reviewer: Not really. Chained operations can be simply put inside `apply`. Object as the return value is useful for returning the object from the function in which a scoped function has been used.

Reviewer 2: >  Chained operations can be simply put inside apply.

This isn't idiomatic. If there's a chain of operations, both `apply` and `also` should be inside that chain, not outside.

Author: Added another example on using also when returning the context object from a function.

`let` can be used to invoke one or more functions on results of call chains. For example, the following code prints the results of two operations on a collection:

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three", "four", "five")
    val resultList = numbers.map { it.length }.filter { it > 3 }
    println(resultList)    
//sampleEnd
}
```
</div>

With `let`, you can rewrite it:

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three", "four", "five")
    numbers.map { it.length }.filter { it > 3 }.let { 
        println(it)
        // and more function calls if needed
    } 
//sampleEnd
}
```
</div>

If the code block contains a single function with `it` as an argument, you can use the method reference (`::`) instead of the lambda:

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three", "four", "five")
    numbers.map { it.length }.filter { it > 3 }.let(::println)
//sampleEnd
}
```
</div>

> `let` is often used for executing a code block only with non-null values. To perform actions on a non-null object, use the safe call operator `?` on it and call `let` with the actions in its lambda.

Reviewer: This is actually the most common use of `let`, so it should be on top.

Reviewer 2: I believe it became the most common even without us writing this page, so it won't harm to leave the narrative order going from basic usages to advanced. Otherwise `let` might get a reputation of a fancy `if (it != null)` replacement.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun processNonNullString(str: String) {}

fun main() {
//sampleStart
    val str: String? = "Hello"   
    //processNonNullString(str)       // compilation error: str can be null
    val length = str?.let { 
        println("let() called on $it")        

>         processNonNullString(it)      // OK: the function is called only on non-nulls

Reviewer: It's important to emphasize here that inside the lambda you should use `it`, not the original variable. Many people write code like `str?.let { processNonNullString(str) }`, and while this works, we need to discourage it.

Author: Added more clarification to the start of the section **Context object: this or it**. I think it's a better place than the description of a particular function.

        it.length
    }
//sampleEnd
}
```
</div>

> Another case for using `let` is introducing local variables for improving code readability. To define a variable, provide its name as the lambda argument so that it can be used instead of the default `it`.

Reviewer: I believe this is an artificial example. To introduce a variable, you can simply declare it.

Reviewer 2: But then that variable could get broader scope that required, and again break the chain of calls. I agree though that the example below doesn't show it in full.

Author: Broadened the description, replaced the example.

`also` is good for performing some actions that take the context object as an argument. Use `also` for additional actions that don't alter the object, such as logging or printing debug information. Usually, you can remove the calls of `also` from the call chain without breaking the program logic. 

When you see `also` in the code, you can read it as “_and also do the following_”.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three")
    numbers.also {
        println("Adding new element to the list")
    }.add("four")
//sampleEnd
}
```
</div>

## Function Selection

To help you choose the right scope function for your purpose, we provide the table of key differences between them. 

|Function|Object reference|Return value|Is extension function|
|---|---|---|---|
|`let`|`it`|Lambda result|Yes|
|`run`|`this`|Lambda result|Yes|
|`run`|-|Lambda result|No: called without the context object|
|`with`|`this`|Lambda result|No: takes the context object as an argument.|
|`apply`|`this`|Context object|Yes|
|`also`|`it`|Context object|Yes|

> Here is a short guide for choosing scope functions depending on the intended purpose:

Reviewer 2: That section in coding conventions could already have external references to it, so we could remove its content and leave only a link to here.

> Here is a short guide for choosing scope functions depending on the intended purpose:

Author: I'd remove the guidance from **Coding conventions** in this case. The guidance in the new page is shorter (which is better if you know the functions) and there are detailed descriptions on the same page (which is good if you don't know the details or need to refresh).
In **Coding conventions**, I'd leave the heading about scope functions and a short introductory word with the link to **Scope functions**. What do you say?

> Here is a short guide for choosing scope functions depending on the intended purpose:

Reviewer 2: Seems fine.

> Here is a short guide for choosing scope functions depending on the intended purpose:

Author: Replaced the old guidance in **Coding conventions** with the link to the new doc.

```kotlin
import kotlin.random.Random

fun writeToLog(message: String) {
    println("INFO: $message")
}

fun main() {
//sampleStart
    fun getRandomInt(): Int {
        return Random.nextInt(100).also {
            writeToLog("getRandomInt() generated value $it")
        }
    }
    
    val i = getRandomInt()
//sampleEnd
}
```
</div>

Additionally, when you pass the context object as an argument, you can provide a custom name for the context object inside the scope.

<div class="sample" markdown="1" theme="idea">
```kotlin
import kotlin.random.Random

fun writeToLog(message: String) {
    println("INFO: $message")
}

fun main() {
//sampleStart
    fun getRandomInt(): Int {
        return Random.nextInt(100).also { value ->
            writeToLog("getRandomInt() generated value $value")
        }
    }
    
    val i = getRandomInt()
//sampleEnd
}
```
</div>

### Return value

The scope functions differ by the result they return:
* `apply` and `also` return the context object.
* `let`, `run`, and `with` return the lambda result.

These two options let you choose the proper function depending on what you do next in your code.

#### Context object 

> The return value of `apply` and `also` is the context object itself. Hence, they can be used as steps in chains of function calls on the object.  

Reviewer 2: It is not clear, what's different from `let` and `run` that can also be used in the middle of call chains. Perhaps, we can call them "side steps" (not sure, maybe "tee steps", or "branch steps"), because they return the same context object on which we can continue chaining calls.

Author: Updated the description

        println("'with' is called with argument $this")
        println("It contains $size elements")
    }
//sampleEnd
}
```
</div>

Another use case for `with` is introducing a helper object whose properties or functions will be used for calculating a value.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three")
    val firstAndLast = with(numbers) {
        "The first element is ${first()}," +
        " the last element is ${last()}"
    }
    println(firstAndLast)
//sampleEnd
}
```
</div>

### `run`

**The context object** is available as a receiver (`this`). **The return value** is the lambda result.

`run` does the same as `with` but invokes as `let` - as an extension function of the context object.

`run` is useful when your lambda contains both the object initialization and the computation of the return value.

<div class="sample" markdown="1" theme="idea">
```kotlin
data class Person(var name: String, var age: Int = 0, var city: String = "")

fun main() {
//sampleStart
    val isTeenager = Person("Adam").run {
        age = 32
        city = "London"
        age > 12 && age < 20
    }

>     println("Is Adam a teenager? $isTeenager")

Reviewer 2: I can't help but feel that this `run` sample is quite contrived compared to its original use case from [KT-5235](https://youtrack.jetbrains.com/issue/KT-5235). I'm tempted to rewrite it with `apply` for setting the properties + `run` for evaluating the expression, but `apply` has not been introduced at this moment.

Perhaps this one could illustrate it better:
```
    class MultiportService(var url: String, var port: Int) {
        fun prepareRequest(): String = "Default request"
        fun query(request: String): String = "Result for query '$request'"
    }
    
// start
    val service = MultiportService("https://example.kotlinlang.org", 80)
      
    val result = service.run {
        port = 8080
        query(prepareRequest() + ", and some extra")
    }
    // an alternative with 'let' could be shown
// end
    println(result)
```
The idea here is the necessity to call several methods of a context object that depend on each other's result or the state of object they mutate.

Author: Thanks, I've replaced the example with this (with small modifications and the alternative with `let`).

---
type: doc
layout: reference
category: "Syntax"
title: "Scope Functions"
---

# Scope Functions

The Kotlin standard library contains several functions whose sole purpose is to execute a block of code within the context of an object. When you call such a function on an object with a [lambda expression](lambdas.html) provided, it forms a temporary scope. In this scope, you can access the object without its name. Such functions are called _scope functions_. There are five of them: `let`, `run`, `with`, `apply`, and `also`.

Basically, these functions do the same: execute a block of code on an object. What's different is how this object becomes available inside the block and what is the result of the whole expression.

Here's a typical usage of a scope function:

<div class="sample" markdown="1" theme="idea">

> ```kotlin

Reviewer 3: \n

data class Person(var name: String, var age: Int, var city: String) {
    fun moveTo(newCity: String) { city = newCity }
    fun incrementAge() { age++ }
}

fun main() {
//sampleStart
    Person("Alice", 20, "Amsterdam").let {
        println(it)
        it.moveTo("London")
        it.incrementAge()
        println(it)
    }
//sampleEnd
}

> ```

Reviewer 3: \n

            add(2.71)
            add(3.14)
            add(1.0)
        }
        .also { println("Sorting the list") }
        .sort()
//sampleEnd
    println(numberList)
}
```
</div>

They also can be used in return statements of functions returning the context object.

<div class="sample" markdown="1" theme="idea">
```kotlin
import kotlin.random.Random

fun writeToLog(message: String) {
    println("INFO: $message")
}

fun main() {
//sampleStart
    fun getRandomInt(): Int {
        return Random.nextInt(100).also {
            writeToLog("getRandomInt() generated value $it")
        }
    }
    
    val i = getRandomInt()
//sampleEnd
}
```
</div>

#### Lambda result

`let`, `run`, and `with` return the lambda result. So, you can use them when assigning the result to a variable, chaining operations on the result, and so on.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three")
    val resultSize = numbers.run { 
        add("four")
        add("five")
        filter { it.endsWith("e") }.size        
    }
    println("There are $resultSize elements that end with e.")
//sampleEnd
}
```
</div>

> Additionally, you can ignore the return value and use a scope function for creating a temporary scope. 

Reviewer 2: In the sample below the result is not ignored.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three", "four", "five")
    numbers.map { it.length }.filter { it > 3 }.let { 
        println(it)
        // and more function calls if needed
    } 
//sampleEnd
}
```
</div>

If the code block contains a single function with `it` as an argument, you can use the method reference (`::`) instead of the lambda:

<div class="sample" markdown="1" theme="idea">
```kotlin
fun main() {
//sampleStart
    val numbers = mutableListOf("one", "two", "three", "four", "five")
    numbers.map { it.length }.filter { it > 3 }.let(::println)
//sampleEnd
}
```
</div>

`let` is often used for executing a code block only with non-null values. To perform actions on a non-null object, use the safe call operator `?.` on it and call `let` with the actions in its lambda.

<div class="sample" markdown="1" theme="idea">
```kotlin
fun processNonNullString(str: String) {}

fun main() {
//sampleStart
    val str: String? = "Hello"   
    //processNonNullString(str)       // compilation error: str can be null
    val length = str?.let { 
        println("let() called on $it")        

>         processNonNullString(it)      // OK: the function is called only on non-nulls

Reviewer 2: ```suggestion
        processNonNullString(it)      // OK: 'it' is not null inside '?.let { }'
```
Otherwise it's unclear which function was meant: `processNonNullString` or `let`.
